templates/patroni.yml.j2

```diff
@@ -48,11 +48,15 @@ bootstrap:
     loop_wait: 10
     retry_timeout: 10
     maximum_lag_on_failover: 1048576
+    synchronous_mode: true
+    synchronous_mode_strict: false
```

Comment:
Strict mode doesn't seem to work with a single unit while doing rolling ops. We'll need to enable/disable based on the size of the cluster.

Reply:
I think `synchronous_mode_strict` should be `false`.
The reason: we still want to have cluster work when the charm shuts down or restarts standbys.

Reply:
> I think `synchronous_mode_strict` should be `false`. The reason: we still want to have cluster work when the charm shuts down or restarts standbys.

I've been testing the HA suite mostly with strict mode enabled, since I have encountered data loss without it. I will retry to see if the latest changes would be at least somewhat stable without it and report back, but I think that Patroni offers the most guarantees on data consistency with strict mode enabled. @taurus-forever @marceloneppel What do you think the correct way to go would be here?

Reply:
Dragomir commented with me on MM about trying to do a failover on the observer process that we recently added to the charm to fix other issues. It may help. If not, I think we should investigate a little bit deeper one specific situation that I found before synchronous mode was enabled and maybe persists now when strict mode is not enabled and lead to the failures Dragomir commented.

On the Patroni [docs](https://patroni.readthedocs.io/en/latest/replication_modes.html#synchronous-mode) we can see `When it is absolutely necessary to guarantee that each write is stored durably on at least two nodes, enable synchronous_mode_strict in addition to the synchronous_mode. This parameter prevents Patroni from switching off the synchronous replication on the primary when no synchronous standby candidates are available.`.

It’s not clear to me if Patroni turns off the synchronous replication due to no replicas or if it’s due to another issue. The situation that I commented above and happened before turning synchronous mode on was that it seemed that there was no synchronous replica even when we requested synchronous replication to be enabled.

In summary, I think we can research if synchronous replication is really working on synchronous mode without strict mode and fix if there is any issue. I saw some other operators not using strict mode, so it seems to be worth to try to avoid it. One last option would be to enable strict mode (if synchronous replications is working, but for some reason it still loses transactions - I think it should’t happen) and add some code to disable it in the observer (this will increase the complexity of the charm, of course), but this is only after we confirm that synchronous replication is working.